File: src/sql/postgres/PostgresRequest.zig
pub fn writeBind(
    name: []const u8,
    cursor_name: bun.String,
    globalObject: *jsc.JSGlobalObject,
    values_array: JSValue,
    columns_value: JSValue,
    parameter_fields: []const int4,
    result_fields: []const protocol.FieldDescription,
    comptime Context: type,
    writer: protocol.NewWriter(Context),
) !void {
    try writer.write("B");
    const length = try writer.length();

    try writer.String(cursor_name);
    try writer.string(name);

    const len: u32 = @truncate(parameter_fields.len);

    // The number of parameter format codes that follow (denoted C
    // below). This can be zero to indicate that there are no
    // parameters or that the parameters all use the default format
    // (text); or one, in which case the specified format code is
    // applied to all parameters; or it can equal the actual number
    // of parameters.
    try writer.short(len);

    var iter = try QueryBindingIterator.init(values_array, columns_value, globalObject);
    for (0..len) |i| {
        const parameter_field = parameter_fields[i];
        const is_custom_type = std.math.maxInt(short) < parameter_field;
        const tag: types.Tag = if (is_custom_type) .text else @enumFromInt(@as(short, @intCast(parameter_field)));

        const force_text = is_custom_type or (tag.isBinaryFormatSupported() and brk: {
            iter.to(@truncate(i));
            if (try iter.next()) |value| {
                break :brk value.isString();
            }
            if (iter.anyFailed()) {
                return error.InvalidQueryBinding;
            }
            break :brk false;
        });

        if (force_text) {
            // If they pass a value as a string, let's avoid attempting to
            // convert it to the binary representation. This minimizes the room
            // for mistakes on our end, such as stripping the timezone
            // differently than what Postgres does when given a timestamp with
            // timezone.
            try writer.short(0);
            continue;
        }

        try writer.short(
            tag.formatCode(),
        );
    }

    // The number of parameter values that follow (possibly zero). This
    // must match the number of parameters needed by the query.
    try writer.short(len);

    debug("Bind: {f} ({d} args)", .{ bun.fmt.quote(name), len });
    iter.to(0);
    var i: usize = 0;
    while (try iter.next()) |value| : (i += 1) {
        const tag: types.Tag = brk: {
            if (i >= len) {
                // parameter in array but not in parameter_fields
                // this is probably a bug a bug in bun lets return .text here so the server will send a error 08P01
                // with will describe better the error saying exactly how many parameters are missing and are expected
                // Example:
                // SQL error: PostgresError: bind message supplies 0 parameters, but prepared statement "PSELECT * FROM test_table WHERE id=$1 .in$0" requires 1
                // errno: "08P01",
                // code: "ERR_POSTGRES_SERVER_ERROR"
                break :brk .text;
            }
            const parameter_field = parameter_fields[i];
            const is_custom_type = std.math.maxInt(short) < parameter_field;
            break :brk if (is_custom_type) .text else @enumFromInt(@as(short, @intCast(parameter_field)));
        };
        if (value.isEmptyOrUndefinedOrNull()) {
            debug("  -> NULL", .{});
            //  As a special case, -1 indicates a
            // NULL parameter value. No value bytes follow in the NULL case.
            try writer.int4(@bitCast(@as(i32, -1)));
            continue;
        }
        if (comptime bun.Environment.enable_logs) {
            debug("  -> {s}", .{tag.tagName() orelse "(unknown)"});
        }

        switch (
        // If they pass a value as a string, let's avoid attempting to
        // convert it to the binary representation. This minimizes the room
        // for mistakes on our end, such as stripping the timezone
        // differently than what Postgres does when given a timestamp with
        // timezone.
        if (tag.isBinaryFormatSupported() and value.isString()) .text else tag) {
            .jsonb, .json => {
                var str = bun.String.empty;
                defer str.deref();
                // Use jsonStringifyFast for SIMD-optimized serialization
                try value.jsonStringifyFast(globalObject, &str);
                const slice = str.toUTF8WithoutRef(bun.default_allocator);
                defer slice.deinit();
                const l = try writer.length();
                try writer.write(slice.slice());
                try l.writeExcludingSelf();
            },
            .bool => {
                const l = try writer.length();
                try writer.write(&[1]u8{@intFromBool(value.toBoolean())});
                try l.writeExcludingSelf();
            },
            .timestamp, .timestamptz => {
                const l = try writer.length();
                try writer.int8(try types.date.fromJS(globalObject, value));
                try l.writeExcludingSelf();
            },
            .bytea => {
                var bytes: []const u8 = "";
                if (value.asArrayBuffer(globalObject)) |buf| {
                    bytes = buf.byteSlice();
                }
                const l = try writer.length();
                debug("    {d} bytes", .{bytes.len});

                try writer.write(bytes);
                try l.writeExcludingSelf();
            },
            .int4 => {
                const l = try writer.length();
                try writer.int4(@bitCast(try value.coerceToInt32(globalObject)));
                try l.writeExcludingSelf();
            },
            .int4_array => {
                const l = try writer.length();
                try writer.int4(@bitCast(try value.coerceToInt32(globalObject)));
                try l.writeExcludingSelf();
            },
            .float8 => {
                const l = try writer.length();
                try writer.f64(@bitCast(try value.toNumber(globalObject)));
                try l.writeExcludingSelf();
            },

            else => {
                const str = try String.fromJS(value, globalObject);
                if (str.tag == .Dead) return error.OutOfMemory;
                defer str.deref();
                const slice = str.toUTF8WithoutRef(bun.default_allocator);
                defer slice.deinit();
                const l = try writer.length();
                try writer.write(slice.slice());
                try l.writeExcludingSelf();
            },
        }
    }

    var any_non_text_fields: bool = false;
    for (result_fields) |field| {
        if (field.typeTag().isBinaryFormatSupported()) {
            any_non_text_fields = true;
            break;
        }
    }

    if (any_non_text_fields) {
        try writer.short(result_fields.len);
        for (result_fields) |field| {
            try writer.short(
                field.typeTag().formatCode(),
            );
        }
    } else {
        try writer.short(0);
    }

    try length.write();
}

pub fn writeQuery(
    query: []const u8,
    name: []const u8,
    params: []const int4,
    comptime Context: type,
    writer: protocol.NewWriter(Context),
) AnyPostgresError!void {
    {
        var q = protocol.Parse{
            .name = name,
            .params = params,
            .query = query,
        };
        try q.writeInternal(Context, writer);
        debug("Parse: {f}", .{bun.fmt.quote(query)});
    }

    {
        var d = protocol.Describe{
            .p = .{
                .prepared_statement = name,
            },
        };
        try d.writeInternal(Context, writer);
        debug("Describe: {f}", .{bun.fmt.quote(name)});
    }
}

pub fn prepareAndQueryWithSignature(
    globalObject: *jsc.JSGlobalObject,
    query: []const u8,
    array_value: JSValue,
    comptime Context: type,
    writer: protocol.NewWriter(Context),
    signature: *Signature,
) AnyPostgresError!void {
    try writeQuery(query, signature.prepared_statement_name, signature.fields, Context, writer);
    try writeBind(signature.prepared_statement_name, bun.String.empty, globalObject, array_value, .zero, &.{}, &.{}, Context, writer);
    var exec = protocol.Execute{
        .p = .{
            .prepared_statement = signature.prepared_statement_name,
        },
    };
    try exec.writeInternal(Context, writer);

    try writer.write(&protocol.Flush);
    try writer.write(&protocol.Sync);
}

pub fn bindAndExecute(
    globalObject: *jsc.JSGlobalObject,
    statement: *PostgresSQLStatement,
    array_value: JSValue,
    columns_value: JSValue,
    comptime Context: type,
    writer: protocol.NewWriter(Context),
) !void {
    try writeBind(statement.signature.prepared_statement_name, bun.String.empty, globalObject, array_value, columns_value, statement.parameters, statement.fields, Context, writer);
    var exec = protocol.Execute{
        .p = .{
            .prepared_statement = statement.signature.prepared_statement_name,
        },
    };
    try exec.writeInternal(Context, writer);

    try writer.write(&protocol.Flush);
    try writer.write(&protocol.Sync);
}

pub fn executeQuery(
    query: []const u8,
    comptime Context: type,
    writer: protocol.NewWriter(Context),
) !void {
    try protocol.writeQuery(query, Context, writer);
    try writer.write(&protocol.Flush);
    try writer.write(&protocol.Sync);
}

pub fn onData(
    connection: *PostgresSQLConnection,
    comptime Context: type,
    reader: protocol.NewReader(Context),
) !void {
    while (true) {
        reader.markMessageStart();
        const c = try reader.int(u8);
        debug("read: {c}", .{c});
        switch (c) {
            'D' => try connection.on(.DataRow, Context, reader),
            'd' => try connection.on(.CopyData, Context, reader),
            'S' => {
                if (connection.tls_status == .message_sent) {
                    bun.debugAssert(connection.tls_status.message_sent == 8);
                    connection.tls_status = .ssl_ok;
                    connection.setupTLS();
                    return;
                }

                try connection.on(.ParameterStatus, Context, reader);
            },
            'Z' => try connection.on(.ReadyForQuery, Context, reader),
            'C' => try connection.on(.CommandComplete, Context, reader),
            '2' => try connection.on(.BindComplete, Context, reader),
            '1' => try connection.on(.ParseComplete, Context, reader),
            't' => try connection.on(.ParameterDescription, Context, reader),
            'T' => try connection.on(.RowDescription, Context, reader),
            'R' => try connection.on(.Authentication, Context, reader),
            'n' => try connection.on(.NoData, Context, reader),
            'K' => try connection.on(.BackendKeyData, Context, reader),
            'E' => try connection.on(.ErrorResponse, Context, reader),
            's' => try connection.on(.PortalSuspended, Context, reader),
            '3' => try connection.on(.CloseComplete, Context, reader),
            'G' => try connection.on(.CopyInResponse, Context, reader),
            'N' => {
                if (connection.tls_status == .message_sent) {
                    connection.tls_status = .ssl_not_available;
                    debug("Server does not support SSL", .{});
                    if (connection.ssl_mode == .require) {
                        connection.fail("Server does not support SSL", error.TLSNotAvailable);
                        return;
                    }
                    continue;
                }

                try connection.on(.NoticeResponse, Context, reader);
            },
            'I' => try connection.on(.EmptyQueryResponse, Context, reader),
            'H' => try connection.on(.CopyOutResponse, Context, reader),
            'c' => try connection.on(.CopyDone, Context, reader),
            'W' => try connection.on(.CopyBothResponse, Context, reader),

            else => {
                debug("Unknown message: {c}", .{c});
                const to_skip = try reader.length() -| 1;
                debug("to_skip: {d}", .{to_skip});
                try reader.skip(@intCast(@max(to_skip, 0)));
            },
        }
    }
}

pub const Queue = bun.LinearFifo(*PostgresSQLQuery, .Dynamic);

const debug = bun.Output.scoped(.Postgres, .visible);

const PostgresSQLConnection = @import("./PostgresSQLConnection.zig");
const PostgresSQLQuery = @import("./PostgresSQLQuery.zig");
const PostgresSQLStatement = @import("./PostgresSQLStatement.zig");
const Signature = @import("./Signature.zig");
const protocol = @import("./PostgresProtocol.zig");
const std = @import("std");
const QueryBindingIterator = @import("../shared/QueryBindingIterator.zig").QueryBindingIterator;

const types = @import("./PostgresTypes.zig");
const AnyPostgresError = @import("./PostgresTypes.zig").AnyPostgresError;
const int4 = types.int4;
const short = types.short;

const bun = @import("bun");
const String = bun.String;

const jsc = bun.jsc;
const JSValue = jsc.JSValue;

File: src/sql_jsc/postgres/PostgresRequest.rs
use crate::jsc::{ErrorCode, JSGlobalObject, JSValue, StringJsc as _};
use bun_core::String as BunString;
use bun_core::fmt as bun_fmt;

use bun_sql::postgres::PostgresProtocol as protocol;
use bun_sql::postgres::PostgresTypes as types;
use bun_sql::postgres::PostgresTypes::{AnyPostgresError, Int4, Short};
use bun_sql::postgres::Status;
use bun_sql::postgres::protocol::{ReaderContext, WriterContext};

use crate::jsc::js_error_to_postgres;
use crate::postgres::PostgresSQLConnection;
use crate::postgres::PostgresSQLQuery;
use crate::postgres::PostgresSQLStatement;
use crate::postgres::Signature;
use crate::shared::QueryBindingIterator;

bun_core::declare_scope!(Postgres, visible);

/// The set of backend message tags `PostgresSQLConnection.on()` dispatches over. Defined here
/// (the dispatch site) rather than in `bun_sql::postgres::protocol` because
/// it is purely a dispatch tag with no wire encoding.
#[derive(Clone, Copy, PartialEq, Eq, Debug, strum::IntoStaticStr)]
pub enum MessageType {
    DataRow,
    CopyData,
    ParameterStatus,
    ReadyForQuery,
    CommandComplete,
    BindComplete,
    ParseComplete,
    ParameterDescription,
    RowDescription,
    Authentication,
    NoData,
    BackendKeyData,
    ErrorResponse,
    PortalSuspended,
    CloseComplete,
    CopyInResponse,
    NoticeResponse,
    EmptyQueryResponse,
    CopyOutResponse,
    CopyDone,
    CopyBothResponse,
    NotificationResponse,
}

/// The PostgreSQL wire protocol uses 16-bit integers for parameter and column counts.
const MAX_PARAMETERS: usize = u16::MAX as usize;

/// The type the server declared for a parameter (ParameterDescription), or the
/// one `Signature` declared when Bind is written before the server answers.
/// OIDs outside the built-in range are user-defined types and bind as text.
fn parameter_tag(oid: Int4) -> types::Tag {
    Short::try_from(oid).map_or(types::Tag::text, types::Tag)
}

/// The binary Bind encodings `write_bind` implements.
#[derive(Clone, Copy)]
enum BinaryEncoding {
    Bool,
    Timestamp,
    Bytea,
    Int4,
    Float8,
}

/// A parameter is sent in the binary format only when the JS value has an
/// exact binary encoding for the declared type. Every other value is sent as
/// text for the server to parse or reject, as string parameters always are.
fn binary_encoding(tag: types::Tag, value: JSValue) -> Option<BinaryEncoding> {
    let encoding = match tag {
        types::Tag::bool if value.is_boolean() => BinaryEncoding::Bool,
        types::Tag::timestamp | types::Tag::timestamptz if value.is_date() || value.is_number() => {
            BinaryEncoding::Timestamp
        }
        types::Tag::bytea if value.is_cell() && value.js_type().is_array_buffer_like() => {
            BinaryEncoding::Bytea
        }
        types::Tag::int4 if is_int4(value) => BinaryEncoding::Int4,
        types::Tag::float8 if value.is_number() => BinaryEncoding::Float8,
        _ => return None,
    };
    Some(encoding)
}

/// A number that `int4` holds exactly.
fn is_int4(value: JSValue) -> bool {
    if !value.is_number() {
        return false;
    }
    let number = value.as_number();
    number.trunc() == number && number >= f64::from(i32::MIN) && number <= f64::from(i32::MAX)
}

pub(crate) fn write_bind<Context: WriterContext>(
    name: &[u8],
    cursor_name: &BunString,
    global: &JSGlobalObject,
    values_array: JSValue,
    columns_value: JSValue,
    parameter_fields: &[Int4],
    result_fields: &[protocol::FieldDescription],
    writer: protocol::NewWriter<Context>,
) -> Result<(), AnyPostgresError> {
    writer.write(b"B")?;
    let length = writer.length()?;

    // The bun.String overload is `bun_string` on NewWriter.
    writer.bun_string(cursor_name)?;
    writer.string(name)?;

    if parameter_fields.len() > MAX_PARAMETERS {
        return Err(AnyPostgresError::TooManyParameters);
    }

    let len: u16 = u16::try_from(parameter_fields.len()).expect("int cast");

    // The number of parameter format codes that follow (denoted C
    // below). This can be zero to indicate that there are no
    // parameters or that the parameters all use the default format
    // (text); or one, in which case the specified format code is
    // applied to all parameters; or it can equal the actual number
    // of parameters.
    writer.short(len)?;

    let mut iter = QueryBindingIterator::init(values_array, columns_value, global)
        .map_err(js_error_to_postgres)?;
    for (i, &oid) in parameter_fields.iter().enumerate() {
        iter.to(i as u32);
        let value = match iter.next().map_err(js_error_to_postgres)? {
            Some(value) => value,
            None if iter.any_failed() => return Err(AnyPostgresError::InvalidQueryBinding),
            None => JSValue::UNDEFINED,
        };
        let format_code: u16 = match binary_encoding(parameter_tag(oid), value) {
            Some(_) => 1,
            None => 0,
        };
        writer.short(format_code)?;
    }

    // The number of parameter values that follow (possibly zero). This
    // must match the number of parameters needed by the query.
    writer.short(len)?;

    bun_core::scoped_log!(Postgres, "Bind: {} ({} args)", bun_fmt::quote(name), len);
    iter.to(0);
    let mut i: usize = 0;
    while let Some(value) = iter.next().map_err(js_error_to_postgres)? {
        // A value with no declared parameter is written as text. The server then
        // rejects the Bind with 08P01 and names both counts, which is a better
        // error than anything that could be produced here.
        let tag = parameter_fields
            .get(i)
            .map_or(types::Tag::text, |&oid| parameter_tag(oid));
        i += 1;
        if value.is_empty_or_undefined_or_null() {
            bun_core::scoped_log!(Postgres, "  -> NULL");
            //  As a special case, -1 indicates a
            // NULL parameter value. No value bytes follow in the NULL case.
            writer.int4((-1i32) as u32)?;
            continue;
        }
        bun_core::scoped_log!(Postgres, "  -> {}", tag.tag_name().unwrap_or("(unknown)"));

        let l = writer.length()?;
        match binary_encoding(tag, value) {
            Some(BinaryEncoding::Bool) => writer.write(&[u8::from(value.as_boolean())])?,
            Some(BinaryEncoding::Timestamp) => writer.int8(
                crate::postgres::types::date::from_js(global, value)
                    .map_err(js_error_to_postgres)?,
            )?,
            Some(BinaryEncoding::Bytea) => {
                if let Some(buffer) = value.as_array_buffer(global) {
                    bun_core::scoped_log!(Postgres, "    {} bytes", buffer.byte_slice().len());
                    writer.write(buffer.byte_slice())?;
                }
            }
            Some(BinaryEncoding::Int4) => writer.int4(value.as_number() as i32 as u32)?,
            Some(BinaryEncoding::Float8) => writer.f64(value.as_number())?,
            None => match tag {
                types::Tag::jsonb | types::Tag::json => {
                    // Use jsonStringifyFast for SIMD-optimized serialization
                    let str = value
                        .json_stringify_fast(global)
                        .map_err(js_error_to_postgres)?;
                    writer.write(str.to_utf8().slice())?;
                }
                // Text input for bytea takes any character sequence, so a value
                // that is neither bytes nor a string has no encoding the server
                // could reject.
                types::Tag::bytea if !value.is_string() => {
                    let received = JSGlobalObject::determine_specific_type(global, value)
                        .map_err(js_error_to_postgres)?;
                    return Err(js_error_to_postgres(
                        global
                            .err(
                                ErrorCode::INVALID_ARG_TYPE,
                                format_args!(
                                    "A bytea parameter must be a string, Buffer, ArrayBuffer or TypedArray. Received {received}"
                                ),
                            )
                            .throw(),
                    ));
                }
                _ if value.is_date() => {
                    let mut buf = [0u8; 64];
                    let Some(iso) = value.to_iso_string(global, &mut buf) else {
                        return Err(js_error_to_postgres(
                            crate::postgres::types::date::throw_invalid_date(global),
                        ));
                    };
                    writer.write(iso)?;
                }
                _ => {
                    let str = BunString::from_js(value, global).map_err(js_error_to_postgres)?;
                    if str.tag() == bun_core::Tag::Dead {
                        return Err(AnyPostgresError::OutOfMemory);
                    }
                    writer.write(str.to_utf8().slice())?;
                }
            },
        }
        l.write_excluding_self()?;
    }

    let mut any_non_text_fields: bool = false;
    for field in result_fields {
        if field.type_tag().is_binary_format_supported() {
            any_non_text_fields = true;
            break;
        }
    }

    if any_non_text_fields {
        if result_fields.len() > MAX_PARAMETERS {
            return Err(AnyPostgresError::TooManyParameters);
        }
        writer.short(result_fields.len())?;
        for field in result_fields {
            writer.short(field.type_tag().format_code())?;
        }
    } else {
        writer.short(0)?;
    }

    length.write()?;
    Ok(())
}

pub(crate) fn write_query<Context: WriterContext>(
    query: &[u8],
    name: &[u8],
    params: &[Int4],
    mut writer: protocol::NewWriter<Context>,
) -> Result<(), AnyPostgresError> {
    {
        let q = protocol::Parse {
            name,
            params,
            query,
        };
        q.write_internal(&mut writer)?;
        bun_core::scoped_log!(Postgres, "Parse: {}", bun_fmt::quote(query));
    }

    {
        let d = protocol::Describe {
            p: protocol::PortalOrPreparedStatement::PreparedStatement(name),
        };
        d.write_internal(writer)?;
        bun_core::scoped_log!(Postgres, "Describe: {}", bun_fmt::quote(name));
    }

    Ok(())
}

pub(crate) fn prepare_and_query_with_signature<Context: WriterContext>(
    global: &JSGlobalObject,
    query: &[u8],
    array_value: JSValue,
    writer: protocol::NewWriter<Context>,
    signature: &mut Signature,
) -> Result<(), AnyPostgresError> {
    writer.atomically(|mut writer| {
        write_query(
            query,
            &signature.prepared_statement_name,
            &signature.fields,
            writer,
        )?;
        write_bind(
            &signature.prepared_statement_name,
            &BunString::EMPTY,
            global,
            array_value,
            JSValue::ZERO,
            &[],
            &[],
            writer,
        )?;
        let exec = protocol::Execute {
            p: protocol::PortalOrPreparedStatement::PreparedStatement(
                &signature.prepared_statement_name,
            ),
            ..Default::default()
        };
        exec.write_internal(&mut writer)?;

        writer.write(&protocol::FLUSH)?;
        writer.write(&protocol::SYNC)?;
        Ok(())
    })
}

pub(crate) fn bind_and_execute<Context: WriterContext>(
    global: &JSGlobalObject,
    statement: &PostgresSQLStatement,
    array_value: JSValue,
    columns_value: JSValue,
    writer: protocol::NewWriter<Context>,
) -> Result<(), AnyPostgresError> {
    writer.atomically(|mut writer| {
        write_bind(
            &statement.signature.prepared_statement_name,
            &BunString::EMPTY,
            global,
            array_value,
            columns_value,
            &statement.parameters,
            &statement.fields,
            writer,
        )?;
        let exec = protocol::Execute {
            p: protocol::PortalOrPreparedStatement::PreparedStatement(
                &statement.signature.prepared_statement_name,
            ),
            ..Default::default()
        };
        exec.write_internal(&mut writer)?;

        writer.write(&protocol::FLUSH)?;
        writer.write(&protocol::SYNC)?;
        Ok(())
    })
}

/// Atomically sends Parse + [Describe] + Bind + Execute + Flush + Sync as a single message batch.
/// This is required for unnamed prepared statements to work correctly with connection poolers
/// like PgBouncer in transaction mode, which may reassign server connections between protocol
/// round-trips. Without this, Parse and Bind+Execute could be routed to different backend
/// connections, causing queries to execute against the wrong prepared statement.
pub(crate) fn parse_and_bind_and_execute<Context: WriterContext>(
    global: &JSGlobalObject,
    query: &[u8],
    statement: &PostgresSQLStatement,
    array_value: JSValue,
    columns_value: JSValue,
    include_describe: bool,
    writer: protocol::NewWriter<Context>,
) -> Result<(), AnyPostgresError> {
    let name = &statement.signature.prepared_statement_name;

    writer.atomically(|mut writer| {
        // Parse
        {
            let q = protocol::Parse {
                name,
                params: &statement.signature.fields,
                query,
            };
            q.write_internal(&mut writer)?;
            bun_core::scoped_log!(Postgres, "Parse: {}", bun_fmt::quote(query));
        }

        // Describe (needed on first execution to learn parameter/result types for caching)
        if include_describe {
            let d = protocol::Describe {
                p: protocol::PortalOrPreparedStatement::PreparedStatement(name),
            };
            d.write_internal(writer)?;
            bun_core::scoped_log!(Postgres, "Describe: {}", bun_fmt::quote(name));
        }

        // Bind — use server-provided types if available (binary format), otherwise
        // fall back to signature types (text format for unknowns). The server will
        // handle text-to-type conversion based on the parameter types from Parse.
        let param_fields = if !statement.parameters.is_empty() {
            &statement.parameters[..]
        } else {
            &statement.signature.fields[..]
        };
        let result_fields = &statement.fields;

        write_bind(
            name,
            &BunString::EMPTY,
            global,
            array_value,
            columns_value,
            param_fields,
            result_fields,
            writer,
        )?;

        // Execute
        let exec = protocol::Execute {
            p: protocol::PortalOrPreparedStatement::PreparedStatement(name),
            ..Default::default()
        };
        exec.write_internal(&mut writer)?;

        writer.write(&protocol::FLUSH)?;
        writer.write(&protocol::SYNC)?;
        Ok(())
    })
}

pub(crate) fn execute_query<Context: WriterContext>(
    query: &[u8],
    mut writer: protocol::NewWriter<Context>,
) -> Result<(), AnyPostgresError> {
    // A simple Query ('Q') is its own sync point: the backend always answers it
    // with exactly one ReadyForQuery. Do not append a Sync here: it would elicit
    // a second, unaccounted ReadyForQuery that re-arms advance() mid-prepare.
    protocol::write_query(query, &mut writer)?;
    writer.write(&protocol::FLUSH)?;
    Ok(())
}

pub(crate) fn on_data<Context: ReaderContext>(
    connection: &PostgresSQLConnection,
    mut reader: protocol::NewReader<Context>,
) -> Result<(), AnyPostgresError> {
    use MessageType as M;
    loop {
        // `fail()` inside a handler tears the connection down (status = Failed,
        // socket closed, queue rejected). Stop dispatching: later messages in
        // the same read must not act on the dead connection.
        if connection.status.get() == Status::Failed {
            return Ok(());
        }
        reader.mark_message_start();
        let c = reader.int::<u8>()?;
        bun_core::scoped_log!(Postgres, "read: {}", c as char);

        // The SSLRequest reply is a bare Byte1('S'|'N') with no Int32 length;
        // it is the only unframed backend byte and must be handled before the
        // frame peek below.
        if let TlsStatus::MessageSent(n) = connection.tls_status.get() {
            match c {
                b'S' => {
                    debug_assert!(n == 8);
                    connection.tls_status.set(TlsStatus::SslOk);
                    connection.setup_tls();
                    return Ok(());
                }
                b'N' => {
                    connection.tls_status.set(TlsStatus::SslNotAvailable);
                    bun_core::scoped_log!(Postgres, "Server does not support SSL");
                    if matches!(
                        connection.ssl_mode,
                        SslMode::Require | SslMode::VerifyCa | SslMode::VerifyFull
                    ) {
                        connection.fail(
                            b"Server does not support SSL",
                            AnyPostgresError::TLSNotAvailable,
                        );
                        return Ok(());
                    }
                    continue;
                }
                _ => return Err(AnyPostgresError::UnexpectedMessage),
            }
        }

        // Every other backend message is Byte1(type) Int32(length) body[length-4].
        // Peek the length here (each handler reads it again) so the handler's
        // net consumption can be checked against it: a handler that leaves the
        // cursor anywhere but the next message's type byte has either scanned a
        // string past the frame or returned with tail bytes still in it, and
        // the stream is unrecoverable (libpq: "message contents do not agree
        // with length in message").
        let (before, length) = reader.peek_length()?;
        let after = before - length;

        match c {
            b'D' => connection.on(M::DataRow, reader.reborrow())?,
            b'd' => connection.on(M::CopyData, reader.reborrow())?,
            b'S' => connection.on(M::ParameterStatus, reader.reborrow())?,
            b'Z' => connection.on(M::ReadyForQuery, reader.reborrow())?,
            b'C' => connection.on(M::CommandComplete, reader.reborrow())?,
            b'2' => connection.on(M::BindComplete, reader.reborrow())?,
            b'1' => connection.on(M::ParseComplete, reader.reborrow())?,
            b't' => connection.on(M::ParameterDescription, reader.reborrow())?,
            b'T' => connection.on(M::RowDescription, reader.reborrow())?,
            b'R' => connection.on(M::Authentication, reader.reborrow())?,
            b'n' => connection.on(M::NoData, reader.reborrow())?,
            b'K' => connection.on(M::BackendKeyData, reader.reborrow())?,
            b'E' => connection.on(M::ErrorResponse, reader.reborrow())?,
            b's' => connection.on(M::PortalSuspended, reader.reborrow())?,
            b'3' => connection.on(M::CloseComplete, reader.reborrow())?,
            b'G' => connection.on(M::CopyInResponse, reader.reborrow())?,
            b'N' => connection.on(M::NoticeResponse, reader.reborrow())?,
            b'I' => connection.on(M::EmptyQueryResponse, reader.reborrow())?,
            b'H' => connection.on(M::CopyOutResponse, reader.reborrow())?,
            b'c' => connection.on(M::CopyDone, reader.reborrow())?,
            b'W' => connection.on(M::CopyBothResponse, reader.reborrow())?,
            b'A' => connection.on(M::NotificationResponse, reader.reborrow())?,

            _ => {
                bun_core::scoped_log!(Postgres, "Unknown message: {}", c as char);
                reader.skip_message()?;
            }
        }

        if connection.status.get() == Status::Failed {
            return Ok(());
        }
        if reader.peek().len() != after {
            bun_core::scoped_log!(
                Postgres,
                "message contents do not agree with length ({}): '{}' left {} of {}",
                length,
                c as char,
                reader.peek().len(),
                after,
            );
            return Err(AnyPostgresError::InvalidMessage);
        }
    }
}

/// Each entry holds a ref on its query.
pub(crate) type Queue = std::collections::VecDeque<bun_ptr::RefPtr<PostgresSQLQuery>>;

use crate::postgres::postgres_sql_connection::{SslMode, TlsStatus};
